jinahub/indexers/searcher/HnswlibSearcher/__init__.py

```diff
@@ -0,0 +1,101 @@
+from jina_commons.indexers.dump import import_vectors
+
+
+class HnswlibSearcher(Executor):
+    """Hnswlib powered vector indexer
+
+    For more information about the Hnswlib supported parameters, please consult:
+        - https://github.com/nmslib/hnswlib
+
+    .. note::
+        Hnswlib package dependency is only required at the query time.
+    """
+
+    def __init__(
+            self,
+            default_top_k: int = 10,
+            distance: str = 'cosine',
+            dump_path: Optional[str] = None,
+            default_traversal_paths: List[str] = None,
+            ef_construction: int = 400,
+            ef_query: int = 50,
+            max_connection: int = 64,
+            *args,
+            **kwargs,
+    ):
+        """
+        Initialize an HnswlibSearcher
+
+        :param default_top_k: get tok k vectors
+        :param distance: distance can be 'l2', 'ip', or 'cosine'
+        :param dump_path: the path to load ids and vecs
+        :param traverse_path: traverse path on docs, e.g. ['r'], ['c']
+        :param ef_construction: defines a construction time/accuracy trade-off
+        :param ef_query:  sets the query time accuracy/speed trade-off
+        :param max_connection: defines tha maximum number of outgoing connections in the graph
+        :param args:
+        :param kwargs:
+        """
+        super().__init__(*args, **kwargs)
+        self.default_top_k = default_top_k
+        self.distance = distance
+        self.default_traversal_paths = default_traversal_paths or ['r']
+        self.ef_construction = ef_construction
+        self.ef_query = ef_query
+        self.max_connection = max_connection
+        self.logger = get_logger(self)
+        dump_path = dump_path or kwargs.get('runtime_args', {}).get('dump_path', None)
+        if dump_path is not None:
+            self.logger.info('Start building "HnswlibSearcher" from dump data')
+            ids, vecs = import_vectors(dump_path, str(self.metas.pea_id))
+            self._ids = np.array(list(ids))
+            self._vecs = np.array(list(vecs))
+            num_dim = self._vecs.shape[1]
+            self._indexer = hnswlib.Index(space=self.distance, dim=num_dim)
+            self._indexer.init_index(max_elements=len(self._vecs), ef_construction=self.ef_construction, M=self.max_connection)
+
+            self._doc_id_to_offset = {}
+            self._load_index(self._ids, self._vecs)
+        else:
+            self.logger.warning(
+                'No data loaded in "HnswlibSearcher". Use .rolling_update() to re-initialize it...'
+            )
+
+    def _load_index(self, ids, vecs):
+        for idx, v in enumerate(vecs):
+            self._indexer.add_items(v.astype(np.float32), idx)
+            self._doc_id_to_offset[ids[idx]] = idx
+        self._indexer.set_ef(self.ef_query)
+
+    @requests(on='/search')
+    def search(self, docs: DocumentArray, parameters: Dict, **kwargs):
+        if not hasattr(self, '_indexer'):
```

Comment:
docs is `Optional` , we have to support and be robust to the possibility of getting `None`

Reply:
Got it